File: scala/src/main/scala/bowlinggame.scala
class BowlingGame {
  def score(rolls: List[Int]):Int = {
    score(1, rolls)
  }

  def score(frame: Int, rolls: List[Int]):Int = {
    frame match{
      case f if f > 10 => 0
      case _ => rolls match {
        case roll::tail         if (roll == 10)     => roll          + score(frame+1,tail) + firstRollOf(tail) + secondRollOf(tail)
        case roll1::roll2::tail if(roll1+roll2==10) => roll1 + roll2 + score(frame+1,tail) + firstRollOf(tail)
        case roll1::roll2::tail                     => roll1 + roll2 + score(frame+1,tail)
        case roll::tail                             => roll
        case Nil                                    => 0
      }
    }
  }

  private def firstRollOf(rolls: List[Int]):Int = {
    rolls match {
      case Nil => 0
      case head::_tail => head
    }
  }

  private def secondRollOf(rolls: List[Int]):Int = {
    rolls match {
      case Nil => 0
      case _head::tail => firstRollOf(tail)
    }
  }
}

File: fsharp/bowlinggame.fs
namespace BowlingGame
module BowlingGame =

  let first_roll_of rolls =
    match rolls with
      | head::_tail -> head
      | _ -> 0

  let second_roll_of rolls =
    match rolls with
      | _head::tail -> first_roll_of(tail)
      | _ -> 0

  let rec score_for frame rolls =
    match rolls with
      | _ when frame > 10                      -> 0
      | roll::tail when roll=10                -> roll          + score_for(frame+1)(tail) + first_roll_of(tail) + second_roll_of(tail)
      | roll1::roll2::tail when roll1+roll2=10 -> roll1 + roll2 + score_for(frame+1)(tail) + first_roll_of(tail)
      | roll1::roll2::tail                     -> roll1 + roll2 + score_for(frame+1)(tail)
      | roll::tail                             -> roll
      | _ -> 0

  let score rolls =
    score_for 1 rolls
